images/ceph/set-ceph-debug-level

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The Rook Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+if [ -z "$1" ]; then
+  echo "no debug level passed choose between 0 and 20"
+  exit 1
+fi
+
+CEPH_DEBUG_LEVEL=$1
+CEPH_DEBUG_FLAG=(
+  lockdep
+  context
+  crush
+  mds
+  mds_balancer
+  mds_locker
+  mds_log
+  mds_log_expire
+  mds_migrator
+  buffer
+  timer
+  filer
+  striper
+  objecter
+  rados
+  rbd
+  rbd_mirror
+  rbd_replay
+  journaler
+  objectcacher
+  client
+  osd
+  optracker
+  objclass
+  filestore
+  journal
+  ms
+  mon
+  monc
+  paxos
+  tp
+  auth
+  crypto
+  finisher
+  reserver
+  heartbeatmap
+  perfcounter
+  rgw
+  rgw_sync
+  civetweb
+  javaclient
+  asok
+  throttle
+  refs
+  compressor
+  bluestore
+  bluefs
+  bdev
+  kstore
+  rocksdb
+  leveldb
+  memdb
+  fuse
+  mgr
+  mgrc
+  dpdk
+  eventtrace
+)
+
+#############
+# FUNCTIONS #
+#############
+check() {
+  ok_to_run=1
+  if [[ "$CEPH_DEBUG_LEVEL" =~ ^[0-9]+$ ]]; then
+    if [ "$CEPH_DEBUG_LEVEL" -ge 0 ] && [ "$CEPH_DEBUG_LEVEL" -le 20 ]; then
+      ok_to_run=0
+    fi
+    elif [[ "$CEPH_DEBUG_LEVEL" == "default" ]]; then
+    ok_to_run=0
+  fi
+}
+
+exec_ceph_command() {
+  local debug_level=$1
+  local action=set
+  if [[ "$debug_level" == "default" ]]; then
+    action="rm"
+  fi
+  
+  # exec command
+  for flag in "${CEPH_DEBUG_FLAG[@]}"; do
+    ARGS=("$action" global debug_"$flag")
+    if [[ "$debug_level" != "default" ]]; then
+      ARGS+=("$debug_level")
+    fi
+    # put stdout in /dev/null since increase debug log will overflow the terminal
+    echo "ceph config ${ARGS[*]}"
+    ceph config "${ARGS[@]}" &> /dev/null & pids+=($!)
+    
+  done
+  wait "${pids[@]}"
```

Comment:
When I tried it, this has the appearance of hanging for several minutes until it finally completed. Why not just run all the commands one at a time? Then it's not confusing for the user who thinks it's hung.

Reply:
It also makes my toolbox unresponsive, with ~60 processes all trying to complete the configuration.

Reply:
Strange, I don't have that behavior, hanging a few seconds for me. Agreed that it can be confusing for users. We can remove the `wait` and still run the commands in //. It's just that we will get no insurance that they were successfully applied since the script will return immediately and the connection closed.

Reply:
Why not run the commands serially? Then the user will see if there are any errors. It might take a minute or two, but what's the rush? :)

Reply:
Because it's just better to go in // there is no reason to wait for all the commands to go one by one. Let's discuss again :)

Reply:
It's probably just the low powered minikube that was impacted, but definitely curious if anybody else see that behavior as well.

Reply:
I agree that it has the appearance of hanging while waiting for pids to finish. I think it would be good to print a message to users that the script executed in parallel and is waiting for the pids to finish.

Reply:
On my platform when I set level 1:

```
[root@rook-ceph-tools-6b4889fdfd-7kgvd /]# time set-ceph-debug-level 1
ceph config set global debug_lockdep 1
ceph config set global debug_context 1
ceph config set global debug_crush 1
ceph config set global debug_mds 1
ceph config set global debug_mds_balancer 1
ceph config set global debug_mds_locker 1
ceph config set global debug_mds_log 1
ceph config set global debug_mds_log_expire 1
ceph config set global debug_mds_migrator 1
ceph config set global debug_buffer 1
ceph config set global debug_timer 1
ceph config set global debug_filer 1
ceph config set global debug_striper 1
ceph config set global debug_objecter 1
ceph config set global debug_rados 1
ceph config set global debug_rbd 1
ceph config set global debug_rbd_mirror 1
ceph config set global debug_rbd_replay 1
ceph config set global debug_journaler 1
ceph config set global debug_objectcacher 1
ceph config set global debug_client 1
ceph config set global debug_osd 1
ceph config set global debug_optracker 1
ceph config set global debug_objclass 1
ceph config set global debug_filestore 1
ceph config set global debug_journal 1
ceph config set global debug_ms 1
ceph config set global debug_mon 1
ceph config set global debug_monc 1
ceph config set global debug_paxos 1
ceph config set global debug_tp 1
ceph config set global debug_auth 1
ceph config set global debug_crypto 1
ceph config set global debug_finisher 1
ceph config set global debug_reserver 1
ceph config set global debug_heartbeatmap 1
ceph config set global debug_perfcounter 1
ceph config set global debug_rgw 1
ceph config set global debug_rgw_sync 1
ceph config set global debug_civetweb 1
ceph config set global debug_javaclient 1
ceph config set global debug_asok 1
ceph config set global debug_throttle 1
ceph config set global debug_refs 1
ceph config set global debug_compressor 1
ceph config set global debug_bluestore 1
ceph config set global debug_bluefs 1
ceph config set global debug_bdev 1
ceph config set global debug_kstore 1
ceph config set global debug_rocksdb 1
ceph config set global debug_leveldb 1
ceph config set global debug_memdb 1
ceph config set global debug_fuse 1
ceph config set global debug_mgr 1
ceph config set global debug_mgrc 1
ceph config set global debug_dpdk 1
ceph config set global debug_eventtrace 1
waiting for all the new logging configuration to be applied, this can take a few seconds

real    0m4.738s
user    0m25.468s
sys     0m2.079s
```
When I reset to the default:

```
[root@rook-ceph-tools-6b4889fdfd-7kgvd /]# time set-ceph-debug-level default
ceph config rm global debug_lockdep
ceph config rm global debug_context
ceph config rm global debug_crush
ceph config rm global debug_mds
ceph config rm global debug_mds_balancer
ceph config rm global debug_mds_locker
ceph config rm global debug_mds_log
ceph config rm global debug_mds_log_expire
ceph config rm global debug_mds_migrator
ceph config rm global debug_buffer
ceph config rm global debug_timer
ceph config rm global debug_filer
ceph config rm global debug_striper
ceph config rm global debug_objecter
ceph config rm global debug_rados
ceph config rm global debug_rbd
ceph config rm global debug_rbd_mirror
ceph config rm global debug_rbd_replay
ceph config rm global debug_journaler
ceph config rm global debug_objectcacher
ceph config rm global debug_client
ceph config rm global debug_osd
ceph config rm global debug_optracker
ceph config rm global debug_objclass
ceph config rm global debug_filestore
ceph config rm global debug_journal
ceph config rm global debug_ms
ceph config rm global debug_mon
ceph config rm global debug_monc
ceph config rm global debug_paxos
ceph config rm global debug_tp
ceph config rm global debug_auth
ceph config rm global debug_crypto
ceph config rm global debug_finisher
ceph config rm global debug_reserver
ceph config rm global debug_heartbeatmap
ceph config rm global debug_perfcounter
ceph config rm global debug_rgw
ceph config rm global debug_rgw_sync
ceph config rm global debug_civetweb
ceph config rm global debug_javaclient
ceph config rm global debug_asok
ceph config rm global debug_throttle
ceph config rm global debug_refs
ceph config rm global debug_compressor
ceph config rm global debug_bluestore
ceph config rm global debug_bluefs
ceph config rm global debug_bdev
ceph config rm global debug_kstore
ceph config rm global debug_rocksdb
ceph config rm global debug_leveldb
ceph config rm global debug_memdb
ceph config rm global debug_fuse
ceph config rm global debug_mgr
ceph config rm global debug_mgrc
ceph config rm global debug_dpdk
ceph config rm global debug_eventtrace
waiting for all the new logging configuration to be applied, this can take a few seconds

real    0m4.715s
user    0m25.295s
sys     0m2.245s
```

So not happening instantly but a few seconds like 5.

@BlaineEXE added your suggestion for info log.

Reply:
Sounds good. I think when I tested setting to level 10 it took less than 15 seconds on my single-node setup.